mllib/src/main/scala/org/apache/spark/ml/feature/Instance.scala

```diff
@@ -114,6 +133,74 @@ private[spark] object InstanceBlock {
   def blokify(instances: RDD[Instance], blockSize: Int): RDD[InstanceBlock] = {
     instances.mapPartitions(_.grouped(blockSize).map(InstanceBlock.fromInstances))
   }
+
+  def blokifyWithMaxMemUsage(
+      iterator: Iterator[Instance],
+      maxMemUsage: Long): Iterator[InstanceBlock] = {
+    require(maxMemUsage > 0)
+
+    var numCols = -1L
+    val buff = mutable.ArrayBuilder.make[Instance]
+    var buffCnt = 0L
+    var buffNnz = 0L
+    var buffUnitWeight = true
+
+    iterator.flatMap { instance =>
+      if (numCols < 0L) numCols = instance.features.size
+      require(numCols == instance.features.size)
+      val nnz = instance.features.numNonzeros
+      buff += instance
+      buffCnt += 1L
+      buffNnz += nnz
+      buffUnitWeight &&= (instance.weight == 1)
+
+      if (getBlockMemUsage(numCols, buffCnt, buffNnz, buffUnitWeight) >= maxMemUsage) {
+        val block = InstanceBlock.fromInstances(buff.result())
+        buff.clear()
+        buffCnt = 0L
+        buffNnz = 0L
+        buffUnitWeight = true
+        Iterator.single(block)
+      } else Iterator.empty
+    } ++ {
+      if (buffCnt > 0) {
+        val block = InstanceBlock.fromInstances(buff.result())
+        Iterator.single(block)
+      } else Iterator.empty
+    }
+  }
+
+  def blokifyWithMaxMemUsage(
+      instances: RDD[Instance],
+      maxMemUsage: Long): RDD[InstanceBlock] = {
+    require(maxMemUsage > 0)
+    instances.mapPartitions(iter => blokifyWithMaxMemUsage(iter, maxMemUsage))
+  }
+
+
+  /**
+   * Suggested value for BlockSizeInMB, based on performance tests of BLAS operation.
+   *
+   * @param dim size of vector.
+   * @param avgNNZ average nnz of vectors.
+   * @param blasLevel level of BLAS operation.
+   */
+  def inferBlockSizeInMB(
+      dim: Int,
+      avgNNZ: Double,
+      blasLevel: Int = 2): Double = {
+    if (dim <= avgNNZ * 3) {
+      // When the dataset is relatively dense, Spark will use netlib-java for optimised numerical
+      // processing, which will try to use nativeBLAS implementations (like OpenBLAS, Intel MKL),
+      // and fallback to the Java implementation (f2jBLAS) if necessary.
+      // The suggested value for dense cases is 0.25.
+      0.25
+    } else {
+      // When the dataset is sparse, Spark will use its own Scala implementation.
+      // The suggested value for sparse cases is 64.0.
+      64.0
```

Comment:
A little surprise to see the default suddenly jumps from 0.25MB to 64MB. This is very risky because 64MB sparse data could generate much bigger dense result, e.g., in multi-class logistic regression or k-means, if we eventually blockify their implementation. In your benchmark, it seems we start observing speed-up at 1MB. I will be very conservative here.

Reply:
I agree that 64MB will to big for a kmeans with large `k`. For kmeans and multi-class logistic regression,  I added a `blasLevel`, maybe we also need to add a param `k`. But for now we may leave it alone, I agree that we can use a conservative value 1MB here.